File: GitVisual.java
// GitVisual.java
import org.eclipse.jgit.api.Git;
import org.eclipse.jgit.api.LogCommand;
import org.eclipse.jgit.lib.Repository;
import org.eclipse.jgit.lib.Ref;
import org.eclipse.jgit.revwalk.RevCommit;
import org.eclipse.jgit.revwalk.RevWalk;
import org.eclipse.jgit.storage.file.FileRepositoryBuilder;
import java.io.File;
import java.text.SimpleDateFormat;
import java.util.*;

public class GitVisual {
    private static boolean color = true;
    private static boolean graph = true;
    private static boolean all = false;
    private static boolean decorate = true;
    private static boolean oneline = false;
    private static String author = null;
    private static int maxCount = 0;
    private static String repoPath = ".";

    public static void main(String[] args) throws Exception {
        parseArgs(args);
        Repository repo = new FileRepositoryBuilder()
                .setGitDir(new File(repoPath, ".git"))
                .build();
        try (Git git = new Git(repo)) {
            LogCommand log = git.log();
            if (all) log.all();
            // фильтры можно добавить через setSkip
            Iterable<RevCommit> commits = log.call();
            // собираем референсы
            Map<String, List<String>> refs = new HashMap<>();
            if (decorate) {
                for (Ref ref : repo.getAllRefs().values()) {
                    if (ref.getObjectId() != null) {
                        String hash = ref.getObjectId().getName().substring(0, 7);
                        refs.computeIfAbsent(hash, k -> new ArrayList<>()).add(ref.getName());
                    }
                }
            }
            int count = 0;
            for (RevCommit c : commits) {
                if (maxCount > 0 && count >= maxCount) break;
                if (author != null && !c.getAuthorIdent().getName().contains(author)) continue;
                count++;
                boolean last = count == maxCount; // упрощённо
                String connector = (count == 1) ? "└── " : "├── ";
                // в реальности нужно определять последний, но для демонстрации
                String hash = c.getId().getName().substring(0, 7);
                String refStr = "";
                if (decorate && refs.containsKey(hash)) {
                    refStr = " (" + String.join(", ", refs.get(hash)) + ")";
                }
                if (oneline) {
                    System.out.println(connector + colorize(hash, "yellow") + colorize(refStr, "blue") + " " + colorize(c.getShortMessage(), "reset"));
                } else {
                    System.out.println(connector + colorize(hash, "yellow") + colorize(refStr, "blue"));
                    System.out.println("    " + colorize("Author:", "cyan") + " " + c.getAuthorIdent().getName() + " <" + c.getAuthorIdent().getEmailAddress() + ">");
                    SimpleDateFormat df = new SimpleDateFormat("yyyy-MM-dd HH:mm");
                    System.out.println("    " + colorize("Date:", "green") + " " + df.format(c.getAuthorIdent().getWhen()));
                    System.out.println("    " + colorize(c.getShortMessage(), "reset") + "\n");
                }
            }
        }
    }

    private static String colorize(String text, String colorCode) {
        if (!color) return text;
        Map<String, String> codes = new HashMap<>();
        codes.put("red", "\033[91m");
        codes.put("green", "\033[92m");
        codes.put("yellow", "\033[93m");
        codes.put("blue", "\033[94m");
        codes.put("cyan", "\033[96m");
        codes.put("reset", "\033[0m");
        return codes.getOrDefault(colorCode, "") + text + codes.get("reset");
    }

    private static void parseArgs(String[] args) {
        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--path": repoPath = args[++i]; break;
                case "--graph": graph = Boolean.parseBoolean(args[++i]); break;
                case "--all": all = true; break;
                case "--decorate": decorate = Boolean.parseBoolean(args[++i]); break;
                case "--oneline": oneline = true; break;
                case "--author": author = args[++i]; break;
                case "-n": maxCount = Integer.parseInt(args[++i]); break;
                case "--color": color = Boolean.parseBoolean(args[++i]); break;
                case "--help": System.out.println("Usage: ..."); System.exit(0);
            }
        }
    }
}

File: git_visual.rs
// git_visual.rs
use std::env;
use std::path::Path;
use std::process;
use git2::{Repository, Commit, ObjectType, BranchType, Reference, Oid};
use structopt::StructOpt;
use chrono::{DateTime, Local};

#[derive(Debug, StructOpt)]
#[structopt(name = "git_visual")]
struct Opt {
    #[structopt(long, default_value = ".")]
    path: String,

    #[structopt(long, default_value = "true")]
    graph: bool,

    #[structopt(long)]
    all: bool,

    #[structopt(long, default_value = "true")]
    decorate: bool,

    #[structopt(long)]
    oneline: bool,

    #[structopt(long)]
    author: Option<String>,

    #[structopt(long)]
    since: Option<String>,

    #[structopt(long)]
    until: Option<String>,

    #[structopt(short = "n", long)]
    max_count: Option<usize>,

    #[structopt(long, default_value = "true")]
    color: bool,
}

fn colorize(text: &str, color: &str, enabled: bool) -> String {
    if !enabled {
        return text.to_string();
    }
    let code = match color {
        "red" => "\x1b[91m",
        "green" => "\x1b[92m",
        "yellow" => "\x1b[93m",
        "blue" => "\x1b[94m",
        "cyan" => "\x1b[96m",
        _ => "\x1b[0m",
    };
    format!("{}{}\x1b[0m", code, text)
}

fn main() {
    let opt = Opt::from_args();
    let repo = match Repository::open(&opt.path) {
        Ok(r) => r,
        Err(e) => {
            eprintln!("Error opening repo: {}", e);
            process::exit(1);
        }
    };

    let mut revwalk = repo.revwalk().unwrap();
    if opt.all {
        revwalk.push_glob("refs/*").unwrap();
    } else {
        let head = repo.head().unwrap();
        revwalk.push(head.target().unwrap()).unwrap();
    }
    revwalk.set_sorting(git2::Sort::TIME).unwrap();

    let commits: Vec<Commit> = revwalk
        .take(opt.max_count.unwrap_or(usize::MAX))
        .map(|oid| repo.find_commit(oid.unwrap()).unwrap())
        .collect();

    // Собираем референсы для декора
    let mut refs_map: std::collections::HashMap<String, Vec<String>> = std::collections::HashMap::new();
    if opt.decorate {
        // Ветки
        let branches = repo.branches(Some(BranchType::Local)).unwrap();
        for branch in branches {
            let (branch, _) = branch.unwrap();
            let name = branch.name().unwrap().unwrap().to_string();
            let oid = branch.get().target().unwrap();
            let short = oid.to_string()[0..7].to_string();
            refs_map.entry(short).or_insert(Vec::new()).push(name);
        }
        // Теги
        let tags = repo.tag_names(None).unwrap();
        for tag_name in tags.iter().flatten() {
            if let Ok(oid) = repo.refname_to_id(&format!("refs/tags/{}", tag_name)) {
                let short = oid.to_string()[0..7].to_string();
                refs_map.entry(short).or_insert(Vec::new()).push(tag_name.to_string());
            }
        }
        // HEAD
        if let Ok(head) = repo.head() {
            if let Some(oid) = head.target() {
                let short = oid.to_string()[0..7].to_string();
                let name = if head.is_branch() {
                    format!("HEAD -> {}", head.shorthand().unwrap_or(""))
                } else {
                    "HEAD".to_string()
                };
                refs_map.entry(short).or_insert(Vec::new()).push(name);
            }
        }
    }

    let color = opt.color;

    for (i, commit) in commits.iter().enumerate() {
        let last = i == commits.len() - 1;
        let connector = if last { "└── " } else { "├── " };
        let hash = commit.id().to_string();
        let short_hash = &hash[0..7];
        let author = commit.author();
        let author_name = author.name().unwrap_or("unknown");
        let email = author.email().unwrap_or("");
        let date = DateTime::<Local>::from(commit.time().seconds()).format("%Y-%m-%d %H:%M").to_string();
        let msg = commit.message().unwrap_or("").lines().next().unwrap_or("");

        let ref_str = if opt.decorate {
            if let Some(refs) = refs_map.get(short_hash) {
                format!(" ({})", refs.join(", "))
            } else {
                String::new()
            }
        } else {
            String::new()
        };

        if opt.oneline {
            println!("{}{}{}{}{}",
                connector,
                colorize(short_hash, "yellow", color),
                colorize(&ref_str, "blue", color),
                " ",
                colorize(msg, "reset", color));
        } else {
            println!("{}{}{}{}",
                connector,
                colorize(short_hash, "yellow", color),
                colorize(&ref_str, "blue", color));
            println!("{}    {} {} <{}>",
                "",
                colorize("Author:", "cyan", color),
                author_name,
                email);
            println!("{}    {} {}",
                "",
                colorize("Date:", "green", color),
                date);
            println!("{}    {}\n",
                "",
                colorize(msg, "reset", color));
        }
    }
}
